README.md
# My Dead Collection

## Version
Beta 2
## Description
Ce programme récupère sur le site [libramemoria.com/avis] les avis de décès en fonction des critères de recherche spécifiés par l'utilisateur.
## Utilisation
MyDeadCollectionLarge est optimisé pour les recherches retournant plus de 600 résultats (20 pages de résultats sur le site classique). Pour des recherches précises, sans spécification du département ou de la date (notamment les recherches pour un nom précis), MyDeadCollectionLarge n'est pas indiqué.
1. Ouvrir le fichier **searchOptions.js** avec un bloc-notes et renseigner les critères de recherche de la même façon que sur le formulaire de recherche de libramemoria.com/avis. Enregistrer. **Attention :** le fichier searchOptions a un format différent de la version b.0.
3. Exécuter MyDeadCollectionLarge.bat en veillant à ce que searchOptions.js et worker.exe soient dans le même dossier que l'exécutable.
4. Les résultats sont disponibles au format CSV dans **output.csv** dans le même dossier.

/!\ **Attention :** Si un fichier output.csv existe déjà, il sera écrasé par l'exécution du programme.

ChangeLog.md
# Change log of My Dead Collection

## Coming next
* Adding resume after failure feature.
* Feel free to make suggestions!

## vb.2
06/04/2020
* Added automatic recording after "maximum pages" warning
* Corrected a few minor bugs : department handling when specified in searchOptions, last month handling when reaching endDate, process exiting after queue is empty, first day handling if specified in searchOptions, ignoring duplicate values.

## vb.1.2
01/03/2020
* Added "maximum pages" warning to error log (in file)

## vb.1.1
30/03/2020
* [Tech] Fixed "log is undefined" bug in recordResults method.
* Run logs are in console and fatal error logs are in log.log.

## vb.1
27/03/2020
* Added unlimited search in new executable : MyDeadCollectionLarge. Old classic search, limited to 600 results (20 pages) is still available.
* Deprecated : searchOptions.js. The file has a new format, old format is deprecated and will generate an error. See new format by opening searchOptions.js. The vb.0 executable still requires the old format.

## vb.0
19/03/2020

recordResults.js
module.exports = function (outputStream, doc, reqOpts, reqQueue) {
	let dpt = reqOpts.dpt;
	let year = reqOpts.year;
	let month = reqOpts.month; // 0-11
	let firstDayOfMonth = reqOpts.firstDayOfMonth;
	let lastDayOfMonth = reqOpts.lastDayOfMonth;
	let page = reqOpts.page;
	let name = reqOpts.name;
	let additional = reqOpts.additional;
	let recording = (name == "");
	let numOfLines = doc.querySelectorAll('div.tableau_liste div[class=ligne]').length;
	let precPrenom = reqOpts.precPrenom;
	let precNom = reqOpts.precNom;
	doc.querySelectorAll('div[class=tableau_liste]').forEach( (table, tIndex, tables) => {
		let date = table.querySelector('div.titre').textContent.replace(/[^0-9\/]*/g, "");
		table.querySelectorAll('div[class=ligne]').forEach( (line, index, lines) => {
			try {
				let result = [];
				let nomCellArray = line.querySelector('div.nom').textContent
					.replace(/\t/g, "")
					.replace(/\n[^a-z]*\n/gi, "\n")
					.replace(/ *\n */g, "\n")
					.replace(/\n+$/g, "")
					.replace(/^\n+/, "")
					.replace(/ *\(.*/g, "")
					.split("\n");
				
				result[0] = nomCellArray[0]; // Prénom
				result[1] = nomCellArray[1]; // Nom
								
				if (recording && result[0] != precPrenom && result[1] != precNom) {
					result[2] = nomCellArray[2] ? nomCellArray[2].replace(/née /,"") : ""; // Nom de jeune fille
					result[3] = ""; // Ville(s)
					line.querySelectorAll('div.ville > a').forEach( (commune) => {
						result[3] += (result[3] ? "," : "") + commune.textContent;
					});
					result[4] = line.querySelector('div.age').textContent.replace(/[\t ]+$/, "").replace(/^[\t ]+/, ""); // Age
					result[5] = line.querySelector('div.titre_journal > a').textContent.replace(/[\t ]+$/, "").replace(/^[\t ]+/, ""); // Journal
					result[6] = date; // Date de publication
					result[7] = line.querySelector('a').href; // Lien
				
					outputStream.write(result.join(";") + "\r\n");
				}
				
				precPrenom = result[0];
				precNom = result[1];
				
				if (tIndex == tables.length-1 && index == lines.length-1) { // end of page
					reqQueue.counter--;
					console.log("Completed " + (reqOpts.additional ? "additional " : "")
						+ "dpt " + dpt
						+ ", year " + year
						+ ", month " + (month+1)
						+ ", page " + page
						+ (recording ? "" : " without new results"));
					if (reqOpts.page < 20 || numOfLines < 30) {
						reqQueue.unshift({
							dpt: dpt,
							year: year,
							month: month,
							firstDayOfMonth: firstDayOfMonth,
							lastDayOfMonth: lastDayOfMonth,
							page: page+1,
							name: recording ? "" : name,
							additional: additional,
							precPrenom: precPrenom,
							precNom: precNom
						});
					} else if (!recording) {
						let warningStr = "Warning: results for dpt " + dpt + ", day " + date + " [French format] are more than 600.\r\nYou should request these parameters again and by prenom's first letter, letter by letter, on the remaining letters.";
						console.log(warningStr);
						console.error(warningStr);
					} else {
						reqQueue.unshift({
							dpt: dpt,
							year: year,
							month: month,
							firstDayOfMonth: firstDayOfMonth,
							lastDayOfMonth: date.substr(0,2),
							page: 1,
							name: result[1]+result[0],
							additional: true
						});
					}
				}

				if (result[1]+result[0] == name) recording = true;
			}
			catch (err) {
				console.log(err);
				console.log(line.outerHTML);
			}
		});
	});
};

MyDeadCollection.js
const https = require('https');
const jsdom = require('jsdom');
const fs = require('fs');

const {JSDOM} = jsdom;

const searchOptions = require(process.cwd()+'/searchOptions.js');
const recordResults = require('./recordResults.js');
const options = require('./connectOptions.js');

const sourceURL = "https://www.libramemoria.com/avis?";

const outputStream = fs.createWriteStream('output.csv');
outputStream.write("Prenom;Nom;Nom JF;Commune;Age;Journal;Date;Lien\r\n");

const DEFAULT_INTERVAL = searchOptions.data.defalutInterval; // ms
const INTERVAL_DELTA = 5; // ms
const ERR_SAMPLE_LENGTH = 50;
const ERR_TOLERANCE = 0.2; // out of 1

let searchURL = sourceURL
	+ "&prenom=" + searchOptions.options.prenom
	+ "&nom=" + searchOptions.options.nom
	+ "&commune=" + searchOptions.options.commune;

let searchDate = new Date(searchOptions.options.debut || searchOptions.data.startDate);
let endDate;
if (searchOptions.options.fin) {
	endDate = new Date(searchOptions.options.fin);
	let today = new Date();
	if (endDate > today) {endDate = today;}
} else {
	endDate = new Date();
}

let reqQueue = [];

for (let dptIterator = 0;
	dptIterator <= (searchOptions.options.departement || searchOptions.options.commune ? 0 : searchOptions.data.departements.length-1);
	dptIterator++) {
	let dpt = searchOptions.data.departements[dptIterator];
	reqQueue["d"+dpt] = [];
	reqQueue["d"+dpt].yearSum = 0
	while (true) {
		let month = searchDate.getMonth(); // 0-11
		let year = searchDate.getFullYear();
		
		// Get the last day of the month
		searchDate.setMonth(month+1);
		searchDate.setDate(0);
		let lastDayOfMonth;
		if (searchDate < endDate) {
			lastDayOfMonth = searchDate.getDate();
			//Increment the date iterator
			searchDate.setDate(1);
			searchDate.setMonth(month+1);
		} else {
			lastDayOfMonth = endDate.getDate();
			searchDate = new Date(searchOptions.options.debut || searchOptions.data.startDate);
			break;
		}
		reqQueue.push({
			dpt: dpt,
			year: year,
			month: month,
			lastDayOfMonth: lastDayOfMonth,
			page: 1
		});
		if (!reqQueue["d"+dpt][year]) {
			reqQueue["d"+dpt][year] = 1;
			reqQueue["d"+dpt].yearSum++
		} else {
			reqQueue["d"+dpt][year]++
		}
	}
}

let interval = DEFAULT_INTERVAL;
let errCounter = [];
for (let i = 0; i < ERR_SAMPLE_LENGTH; i++) {
	errCounter[i] = 0;
}
let qLoop;

let loopFct = () => {
	let reqOpts = reqQueue.shift();
	if (reqOpts) {
		let pageURL = searchURL
			+ "&departement=" + reqOpts.dpt
			+ "&debut=1/" + (reqOpts.month+1) + "/" + reqOpts.year
			+ "&fin=" + reqOpts.lastDayOfMonth + "/" + (reqOpts.month+1) + "/" + reqOpts.year
			+ "&page=" + reqOpts.page;
		https.get(pageURL, options, (pageRes) => {
			let pageData = '';
			pageRes.on('data', (chunk) => {
				pageData += chunk;
			});
			pageRes.on('end', () => {
				const pageDoc = (new JSDOM (pageData)).window.document;
				if (pageDoc.querySelector('p.noresults')) { // If page is empty 
					console.log("Completed dpt " + reqOpts.dpt + ", year " + reqOpts.year + ", month " + (reqOpts.month+1));
					if (--reqQueue["d"+reqOpts.dpt][reqOpts.year] <= 0) {
						console.log("Completed dpt " + reqOpts.dpt + ", year " + reqOpts.year);
						if (--reqQueue["d"+reqOpts.dpt].yearSum <= 0) {
							console.log("Completed dpt " + reqOpts.dpt);
						}
					}
				}
				else {
					recordResults(
						outputStream,
						pageDoc,
						reqOpts.dpt,
						reqOpts.year,
						reqOpts.month+1, // 1-12
						reqOpts.page
					);
					reqQueue.unshift({
						dpt: reqOpts.dpt,
						year: reqOpts.year,
						month: reqOpts.month,
						lastDayOfMonth: reqOpts.lastDayOfMonth,
						page: reqOpts.page+1
					});
				}
				errCounter.shift();
				errCounter.push(0);
			});
		}).on('error', (err) => {
			console.log("HTTP Error on URL: " + pageURL);
			if (err.code == "ECONNRESET") {
				console.log("Rertrying...");
				reqQueue.unshift(reqOpts);
				errCounter.shift();
				errCounter.push(1);
				if (errCounter.reduce((sum, val) => {
					return sum += val;
				}) > ERR_TOLERANCE*ERR_SAMPLE_LENGTH) {
					clearInterval(qLoop);
					qLoop = setInterval(loopFct, interval += 5);
					console.log("Slowed down!");
				}
			} else {
				console.log(err);
			}
		});
	}
}

qLoop = setInterval(loopFct,interval);

MyDeadCollectionLarge.js
const https = require('https');
const jsdom = require('jsdom');
const fs = require('fs');

const {JSDOM} = jsdom;

const searchOptions = require(process.cwd()+'/searchOptions.js');
const options = require('./connectOptions.js');
const recordResults = require('./recordResults.js');

const sourceURL = "https://www.libramemoria.com/avis?";

const outputStream = fs.createWriteStream('output.csv');
outputStream.write("Prenom;Nom;Nom JF;Commune;Age;Journal;Date;Lien\r\n");

const DEFAULT_INTERVAL = searchOptions.data.defaultInterval; // ms
const MAXIMUM_INTERVAL = 100; // ms
const INTERVAL_DELTA = 5; // ms
const ERR_SAMPLE_LENGTH = 50;
const ERR_TOLERANCE = 0.2; // out of 1
const DELAY_WITH_EMPTY_QUEUE = 4000 // ms

let searchURL = sourceURL
	+ "&prenom=" + searchOptions.options.prenom
	+ "&nom=" + searchOptions.options.nom
	+ "&commune=" + searchOptions.options.commune;

let startDate = new Date(searchOptions.options.debut || searchOptions.data.startDate);
let endDate;
if (searchOptions.options.fin) {
	endDate = new Date(searchOptions.options.fin);
	let today = new Date();
	if (endDate > today) {endDate = today;}
} else {
	endDate = new Date();
}

let reqQueue = [];

for (let dptIterator = 0;
	dptIterator <= (searchOptions.options.departement || searchOptions.options.commune ? 0 : searchOptions.data.departements.length-1);
	dptIterator++) {
	let dpt = searchOptions.options.departement || searchOptions.data.departements[dptIterator];
	reqQueue["d"+dpt] = [];
	reqQueue["d"+dpt].yearSum = 0
	let searchDate = new Date(startDate);
	let loop = true;
	while (loop) {
		let firstDayOfMonth = searchDate.getDate(); // In case it is not 1 for the first month
		let month = searchDate.getMonth(); // 0-11
		let year = searchDate.getFullYear();
		
		// Get the last day of the month
		searchDate.setMonth(month+1);
		searchDate.setDate(0);
		let lastDayOfMonth;
		if (searchDate < endDate) {
			lastDayOfMonth = searchDate.getDate();
			//Increment the date iterator
			searchDate.setDate(1);
			searchDate.setMonth(month+1);
		} else {
			lastDayOfMonth = endDate.getDate();
			loop = false;
		}
		reqQueue.push({
			dpt: dpt,
			year: year,
			month: month,
			firstDayOfMonth: firstDayOfMonth,
			lastDayOfMonth: lastDayOfMonth,
			page: 1,
			name: ""
		});
		if (!reqQueue["d"+dpt][year]) {
			reqQueue["d"+dpt][year] = 1;
			reqQueue["d"+dpt].yearSum++
		} else {
			reqQueue["d"+dpt][year]++
		}
	}
}

let interval = DEFAULT_INTERVAL;
let errCounter = [];
for (let i = 0; i < ERR_SAMPLE_LENGTH; i++) {
	errCounter[i] = 0;
}
let qLoop;

let terminate = false;
let delayCounter = 0;
reqQueue.counter = 0;

let loopFct = () => {
	let reqOpts = reqQueue.shift();
	if (reqOpts) {
		delayCounter = 0;
		let pageURL = searchURL
			+ "&departement=" + reqOpts.dpt
			+ "&debut=" + reqOpts.firstDayOfMonth + "/" + (reqOpts.month+1) + "/" + reqOpts.year
			+ "&fin=" + reqOpts.lastDayOfMonth + "/" + (reqOpts.month+1) + "/" + reqOpts.year
			+ "&page=" + reqOpts.page;
		reqQueue.counter++;
		https.get(pageURL, options, (pageRes) => {
			let pageData = '';
			pageRes.on('data', (chunk) => {
				pageData += chunk;
			});
			pageRes.on('end', () => {
				const pageDoc = (new JSDOM (pageData)).window.document;
				if (pageDoc.querySelector('p.noresults')) { // If page is empty
					reqQueue.counter--;
					console.log("Completed dpt " + reqOpts.dpt + ", year " + reqOpts.year + ", month " + (reqOpts.month+1));
					if (--reqQueue["d"+reqOpts.dpt][reqOpts.year] <= 0) {
						console.log("Completed dpt " + reqOpts.dpt + ", year " + reqOpts.year);
						if (--reqQueue["d"+reqOpts.dpt].yearSum <= 0) {
							console.log("Completed dpt " + reqOpts.dpt);
						}
					}
				}
				else {
					recordResults(
						outputStream,
						pageDoc,
						reqOpts,
						reqQueue
					);
				}
				errCounter.shift();
				errCounter.push(0);
			});
		}).on('error', (err) => {
			reqQueue.counter--;
			switch (err.code) {
				case "ECONNRESET":
					if (interval > MAXIMUM_INTERVAL) {
						console.error("HTTP Error on URL: " + pageURL);
						console.error("Your internet connection is probably too slow.\r\nShutting down...");
						process.exit(1);
					}
					console.log("HTTP Error on URL: " + pageURL);
					console.log("Rertrying...");
					reqQueue.unshift(reqOpts);
					errCounter.shift();
					errCounter.push(1);
					if (errCounter.reduce((sum, val) => {
						return sum += val;
					}) > ERR_TOLERANCE*ERR_SAMPLE_LENGTH) {
						clearInterval(qLoop);
						qLoop = setInterval(loopFct, interval += 5);
						console.log("Slowed down!");
					}
					break;
				case "ENOTFOUND":
					console.error("HTTP Error on URL: " + pageURL);
					console.error(err.message);
					console.error("You may not have an internet connection.\r\nShutting down...");
					process.exit(1);
					break;
				default:
					console.log(err);
					console.error(err);
			}
		});
	} else {
		delayCounter++;
		if (delayCounter*interval > DELAY_WITH_EMPTY_QUEUE && reqQueue.counter <= 0) {
			clearInterval(qLoop);
		}
	}
}

qLoop = setInterval(loopFct, interval);

MyDeadCollectionSmall.js
const https = require('https');
const jsdom = require('jsdom');
const fs = require('fs');

const {JSDOM} = jsdom;

var searchOptions = require(process.cwd()+'/searchOptions.js').options;

var sourceURL = "https://www.libramemoria.com/avis?";

for (let key in searchOptions) {
	sourceURL += key + "=" + searchOptions[key] + "&";
}

var options = {
	headers: {
		Host: "www.libramemoria.com",
		Connection: 'keep-alive',
		'Cache-Control': "max-age=0",
		'Upgrade-Insecure-Requests': 1,
		DNT: 1,
		'User-Agent': "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36",
		'Sec-Fetch-Dest': "document",
		Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
		'Sec-Fetch-Site': "cross-site",
		'Sec-Fetch-Mode': "navigate",
		'Sec-Fetch-User': "?1",
		'Accept-Language': "fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7,it;q=0.6,pt;q=0.5"
	}
}

var recordResults = function (doc, page) {
	let output = "";
	doc.querySelectorAll('div.tableau_liste').forEach( (table) => {
		let date = table.querySelector('div.titre').textContent.replace(/[^1-9\/]*/, "");
		table.querySelectorAll('div[class=ligne]').forEach( (line) => {
			try {
				let result = [];
				let nomCellArray = line.querySelector('div.nom').textContent
					.replace(/\t/g, "")
					.replace(/\n[^a-z]*\n/gi, "\n")
					.replace(/ *\n */g, "\n")
					.replace(/\n+$/g, "")
					.replace(/^\n+/, "")
					.replace(/ *\(.*/g, "")
					.split("\n");
				
				result[0] = nomCellArray[0];
				result[1] = nomCellArray[1];
				result[2] = nomCellArray[2] ? nomCellArray[2].replace(/née /,"") : "";
				result[3] = "";
				line.querySelectorAll('div.ville > a').forEach( (commune) => {
					result[3] += (result[3] ? "," : "") + commune.textContent;
				});
				result[4] = line.querySelector('div.age').textContent.replace(/[\t ]+$/, "").replace(/^[\t ]+/, "");
				result[5] = line.querySelector('div.titre_journal').textContent.replace(/[\t ]+$/, "").replace(/^[\t ]+/, "");
				result[6] = date;
				result[7] = line.querySelector('a').href;
				
				output += result.join(";") + "\r\n";
			}
			catch (err) {
				console.log(err);
				console.log(line.outerHTML);
			}
		});
	});

	outputStream.write(output, (err) => {
		if (err) throw err;
		else console.log("Completed page " + page);
	});
}

var outputStream = fs.createWriteStream('output.csv');
outputStream.write("Prenom;Nom;Nom JF;Commune;Age;Journal;Date;Lien\r\n");
		
let k = 0; // Page iterator
let loop = setInterval ( () => { // For each page. Interval is set in order not to send all the HTTP requests at the same time.
	k++;
	let i = k;
	let pageURL = sourceURL + "&page=" + i;
	https.get(pageURL, options, (pageRes) => {
		let pageData = '';
		pageRes.on('data', (chunk) => {
			pageData += chunk;
		});
		pageRes.on('end', () => {
			const pageDoc = (new JSDOM (pageData)).window.document;
			if (pageDoc.querySelector('p.noresults')) { // If page is empty raises the signal (k=0) to stop setInterval
				if (k != 0) {console.log("Last page reached");}
				k=0;
				clearInterval(loop);
			}
			else { recordResults(pageDoc, i); }
		});
	}).on("error", (err) => {
		console.log("Error: " + err.message);
	});
}, 100);

searchOptions.js
/*
	Critères de recherche.
	Ce fichier est nécessaire à l'exécution du programme et doit être dans le même dossier que l'exécutable.
	Ne pas modifier la structure.
	Renseigner les champs "options" comme sur le formulaire de recherche de libramemoria.com/avis, en modifiant uniquement le texte entre guillements (").
	Si un champ n'est pas renseigné, la valeur doit être : "".
	Ne pas modifier les champs "data".
	
	Format prenom : "Commence par". Ex. : "N" va trouver tous les prénoms débutant par N.
	Format nom : "Commence par" sur Nom et Nom de jeune fille. Ex. : "N" va trouver tous les noms et noms de jeune fille débutant par N.
	Format date : "M/J/AAAA" (format anglais), les mois et les jours peuvent être à 1 ou 2 chiffres.
	Format departement : "XX(X)", le code du département à 2 ou 3 chiffres (pour les départements à 1 chiffre, précéder d'un 0).
	Format commune : "XXXXX", le code INSEE à 5 chiffres. S'il est renseigné, departement est ignoré.
	Codes départements disponibles ci-après.
	
	Exemple :
	
	module.exports = {
		options: {
			prenom: "",
			nom: "",
			debut: "1/15/2015",
			fin: "",
			departement: "02",
			commune: ""
		},
		data: /!\ DO NOT MODIFY
	};
*/

module.exports = {
	options: {
		prenom: "",
		nom: "",
		debut: "", // MM/DD/YYYY
		fin: "", // MM/DD/YYYY
		departement: "",
		commune: ""
	},
	data: { // DO NOT MODIFY
		defaultInterval: 10, // ms
		startDate: "1/1/2010", // MM/DD/YYYY
		departements: ["01","02","03","04","05","06","07","08","09","10","11","12","13","14","15","16","17","18","19","21","22","23","24","25","26","27","28","29","2A","2B","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","971","972","973","974","976"]
	}
};

/*
	Liste des départements :
		01-Ain
		02-Aisne
		03-Allier
		04-Alpes-de-Haute-Provence
		05-Hautes-Alpes
		06-Alpes-Maritimes
		07-Ardèche
		08-Ardennes
		09-Ariège
		10-Aube
		11-Aude
		12-Aveyron
		13-Bouches-du-Rhône
		14-Calvados
		15-Cantal
		16-Charente
		17-Charente-Maritime
		18-Cher
		19-Corrèze
		21-Côte-d'Or
		22-Côtes-d'Armor
		23-Creuse
		24-Dordogne
		25-Doubs
		26-Drôme
		27-Eure
		28-Eure-et-Loir
		29-Finistère
		2A-Corse-du-Sud
		2B-Haute-Corse
		30-Gard
		31-Haute-Garonne
		32-Gers
		33-Gironde
		34-Hérault
		35-Ille-et-Vilaine
		36-Indre
		37-Indre-et-Loire
		38-Isère
		39-Jura
		40-Landes
		41-Loir-et-Cher
		42-Loire
		43-Haute-Loire
		44-Loire-Atlantique
		45-Loiret
		46-Lot
		47-Lot-et-Garonne
		48-Lozère
		49-Maine-et-Loire
		50-Manche
		51-Marne
		52-Haute-Marne
		53-Mayenne
		54-Meurthe-et-Moselle
		55-Meuse
		56-Morbihan
		57-Moselle
		58-Nièvre
		59-Nord
		60-Oise
		61-Orne
		62-Pas-de-Calais
		63-Puy-de-Dôme
		64-Pyrénées-Atlantiques
		65-Hautes-Pyrénées
		66-Pyrénées-Orientales
		67-Bas-Rhin
		68-Haut-Rhin
		69-Rhône
		70-Haute-Saône
		71-Saône-et-Loire
		72-Sarthe
		73-Savoie
		74-Haute-Savoie
		75-Paris
		76-Seine-Maritime
		77-Seine-et-Marne
		78-Yvelines
		79-Deux-Sèvres
		80-Somme
		81-Tarn
		82-Tarn-et-Garonne
		83-Var
		84-Vaucluse
		85-Vendée
		86-Vienne
		87-Haute-Vienne
		88-Vosges
		89-Yonne
		90-TerritoiredeBelfort
		91-Essonne
		92-Hauts-de-Seine
		93-Seine-Saint-Denis
		94-Val-de-Marne
		95-Val-d'Oise
		971-Guadeloupe
		972-Martinique
		973-Guyane
		974-LaRéunion
		976-Mayotte
*/